modules/ding_debt_easy/ding_debt_easy.module

```diff
@@ -0,0 +1,908 @@
+      'path' => $path . '/templates',
+    ],
+  ];
+}
+
+/**
+ * Implements hook_ding_provider().
+ */
+function ding_debt_easy_ding_provider() {
+  return [
+    'title' => 'Net easy payment provider',
+    'provides' => [
+      'payment' => [
+        'prefix' => 'payment',
+        'file' => 'includes/easy.payment.inc',
+      ],
+    ],
+  ];
+}
+
+/**
+ * Implements hook_cron().
+ */
+function ding_debt_easy_cron() {
+  _ding_debt_easy_wipe();
+  _ding_debt_easy_retry_payments();
+}
+
+/**
+ * Wipe old local payment records base on configuration.
+ */
+function _ding_debt_easy_wipe() {
+  $interval = _ding_debt_easy_get_wipe_interval();
+
+  // Zero here is wipe out is disabled in configuration.
+  if (0 < $interval) {
+    $limit = DateTimeImmutable::createFromFormat('U', time() - $interval);
+    $limit = $limit->setTimezone(new DateTimeZone(drupal_get_user_timezone()));
+
+    $res = db_delete('ding_debt_easy')
+      ->condition('changed', $limit->getTimestamp(), '<')
+      ->execute();
+
+    watchdog('ding_debt_easy', '%x payments was removed from the local database.', [
+      '%x' => $res,
+    ], WATCHDOG_INFO);
+  }
+}
+
+/**
+ * Retry pending payments.
+ */
+function _ding_debt_easy_retry_payments() {
+  $orders = db_select('ding_debt_easy', 'dbe')
+    ->fields('dbe', ['payment_id'])
+    ->condition('status', DING_DEBT_EASY_STATUS_PENDING)
+    ->orderBy('changed', 'DESC')
+    ->condition('retries', DING_DEBT_EASY_MAX_RETRIES, '<=')
+    ->execute()
+    ->fetchAll();
+
+  foreach ($orders as $order) {
+    // Load local order from database this way to ensure thing are decoded
+    // correctly.
+    $order = _ding_debt_easy_get_payment_local($order->payment_id);
+
+    // Check interval retry (back-off strategy with 6 retries ending 24 hours
+    // from first try, starting at 10 min - 300, 900, 2700 ... 86400 sec.).
+    $limit = $order['changed'] + (int) (300 * pow(3, $order['retries']));
+
+    if (time() > $limit) {
+      // Check order status at the provider.
+      /** @var \DingProviderDebt[] $fees */
+      $fees = ding_provider_invoke('debt', 'list', NULL, TRUE, $order['patron_id'], TRUE);
+      $chargeable = TRUE;
+      foreach ($order['provider_ids'] as $id) {
+        if (!isset($fees[$id])) {
+          // The fee is not found at the provider.
+          _ding_debt_easy_update_status_local($order['payment_id'], DING_DEBT_EASY_STATUS_FAILED);
+          _ding_debt_easy_update_retries_local($order['payment_id'], $order['retries']);
+          _ding_debt_easy_remove_personal_info($order['payment_id']);
+          $chargeable = FALSE;
+          break;
+        }
+
+        if (empty($fees[$id]->paid_date)) {
+          // The fee has not been marked as paid at the provider. So we simply
+          // try to pay them all once more.
+          $status = ding_provider_invoke('debt', 'payment_received', NULL, $order['provider_ids'], $order['order_id'], $order['patron_id']);
```

Comment:
I do not understand. But if you try to pay something that have been payed at FBS... it will simply answer the same, as first time, as FBS don't pay anything, but just mark a fee as paid.

Reply:
I read the code as follows

```
load a patrons fees from FBS
loop over fees registered with an order at Nets
  identify the corresponding fee from FBS
  if the fee has not been payed at FBS
    mark all fees registered with the order as payed at FBS
    exit the loop
```

I find it odd that if one fee in the order has not been payed then we mark all fees as payed by calling `ding_provider_invoke('debt', 'payment_received')`. Marking something as payed which has already been marked as payed previously may not make a difference but I find it quite confusing. I would have expected us to check their individual status and mark them individually as paid.

The problems continue in the following lines, LOC 227-236 where the patron is charged for the fee. What happens if we have an order with two fees: The first fee is not marked as payed but the second is not. With the logic in this implementation it seems if we would still change payment for the full amount. 

I guess one interpretation of all this might be that if marking something as payed multiple times causes problems then `ding_provider_invoke('debt', 'payment_received')` will return `FALSE` and payment for the order will not be charged again.

All in all the current implementation may be well and good but I find it hard to grok.

Reply:
I have tried to explain the "stuff" behind this "strange" way of handling orders. It comes back to that we never what to charge fees from the user if not 100% sure they have been marked as paid first. And we don't care if, something gets marked as paid, but is not charged.